py/server/deephaven/liveness_scope.py

```diff
@@ -27,44 +50,136 @@ def __init__(self, j_scope: jpy.JType):
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        self.close()
+        _pop(self.j_scope)
+        self.release()
+
+    def release(self):
+        """Closes the LivenessScope and releases all the query graph resources.
+
+        Raises:
+            DHError
+        """
+        try:
+            self.j_scope.release()
+            self.j_scope = None
+        except Exception as e:
+            raise DHError(e, message="failed to close the LivenessScope.")
 
     def close(self):
         """Closes the LivenessScope and releases all the query graph resources.
 
         Raises:
             DHError
         """
-        if self.j_scope:
-            try:
-                _JLivenessScopeStack.pop(self.j_scope)
-                self.j_scope.release()
-                self.j_scope = None
-            except Exception as e:
-                raise DHError(e, message="failed to close the LivenessScope.")
+        warn('This function is deprecated, prefer release(). Use cases that rely on this are likely to now fail.',
+             DeprecationWarning, stacklevel=2)
+        _pop(self.j_scope)
+        self.release()
+
+    @contextlib.contextmanager
+    def open(self, release_after_block: bool = True) -> None:
+        """
+        Uses this scope for the duration of the `with` block. The scope defaults to being closed
+        when the block ends, disable by passing release_after_block=False
+
+        Args:
+            release_after_block: True to release the scope when the block ends, False to leave the
+            scope open for reuse.
+
+        Returns: None, to allow changes in the future.
+
+        """
+        _push(self.j_scope)
+        try:
+            yield None
+        finally:
+            _pop(self.j_scope)
+            if release_after_block:
+                self.release()
 
     @property
     def j_object(self) -> jpy.JType:
         return self.j_scope
 
-    def preserve(self, wrapper: JObjectWrapper) -> None:
+    def preserve(self, referent: Union[JObjectWrapper, jpy.JType]) -> None:
         """Preserves a query graph node (usually a Table) to keep it live for the outer scope.
 
         Args:
-            wrapper (JObjectWrapper): a wrapped Java object such as a Table
+            referent (Union[JObjectWrapper, jpy.JType]): an object to preserve in the next outer liveness scope
 
         Raises:
             DHError
         """
+        referent = _unwrap_to_liveness_referent(referent)
         try:
             _JLivenessScopeStack.pop(self.j_scope)
-            _JLivenessScopeStack.peek().manage(wrapper.j_object)
-            _JLivenessScopeStack.push(self.j_scope)
+            _JLivenessScopeStack.peek().manage(_unwrap_to_liveness_referent(referent))
         except Exception as e:
             raise DHError(e, message="failed to preserve a wrapped object in this LivenessScope.")
+        finally:
+            _JLivenessScopeStack.push(self.j_scope)
```

Comment:
The original code was wrong, but this is differently wrong... working on a fix with more comments.